uarray/_backend.py

```diff
@@ -191,7 +186,17 @@ def set_backend(backend, coerce=False, only=False):
     skip_backend: A context manager that allows skipping of backends.
     set_global_backend: Set a single, global backend for a domain.
     """
-    return _SetBackendContext(backend, coerce, only)
+    try:
+        return _set_backend_cache[weakref.ref(backend), coerce, only]
+    except KeyError:
+
+        def del_attr(x):
+            del _set_backend_cache[x, coerce, only]
+
+        ref = weakref.ref(backend, del_attr)
+        ctx = _SetBackendContext(backend, coerce, only)
+        _set_backend_cache[ref, coerce, only] = ctx
```

Comment:
Note that the context object itself holds an owning reference to the backend; so this still leaks references.  I think this is the hardest problem to solve.

Reply:
If the context object exists, we need the reference to the backend anyway.

Reply:
Note also that the problem we're trying to solve is to GC the context, not the backend.

Reply:
I think you're describing the same issue in different words? You want to remove the context from the cache when the backend is collected, but it's existence in the cache prevents it from being collected.

Reply:
We could cache the three `__ua_*__` attributes and release the backend.

Reply:
Maybe the solution is to put the context in the backend's `__dict__` and release all other references to it?

Reply:
ie. simply have a `__ua_cache__[("skip")|("set", coerce, only)]` attribute on the backend?

Reply:
> We could cache the three __ua_*__ attributes and release the backend.

I think this just moves the problem to the attributes.

> Maybe the solution is to put the context in the backend's __dict__ and release all other references to it?

Good idea, I think that solves both problems doesn't it.

Reply:
i.e. the backend owning a circular reference to itself doesn't matter so we  don't need `weakref` any more.

Reply:
Opened #210 to that effect.